File: app/src/main/java/com/mathocosta/learnbasic/activities/ShoppingListActivity.java
package com.mathocosta.learnbasic.activities;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.mathocosta.learnbasic.R;
import com.mathocosta.learnbasic.activities.adapters.ShoppingListRecListAdapter;
import com.mathocosta.learnbasic.models.ShoppingListItem;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collections;

import static androidx.recyclerview.widget.ItemTouchHelper.ACTION_STATE_DRAG;
import static androidx.recyclerview.widget.RecyclerView.Adapter;
import static androidx.recyclerview.widget.RecyclerView.ViewHolder;

public class ShoppingListActivity
        extends AppCompatActivity
        implements ShoppingListRecListAdapter.ItemClickListener {

    private static final int ADD_ITEM = 1;
    private static final int EDIT_ITEM = 2;

    private ShoppingListRecListAdapter recViewAdapter;
    private ArrayList<ShoppingListItem> shoppingListItems = new ArrayList<ShoppingListItem>();

    private final ItemTouchHelper.SimpleCallback simpleTouchCallback = new ItemTouchHelper.SimpleCallback(
            ItemTouchHelper.DOWN | ItemTouchHelper.UP,
            0
    ) {
        @Override
        public boolean onMove(
                @NonNull RecyclerView recyclerView,
                @NonNull ViewHolder viewHolder,
                @NonNull ViewHolder target
        ) {
            Adapter adapter = recyclerView.getAdapter();
            int fromPosition = viewHolder.getAdapterPosition();
            int toPosition = target.getAdapterPosition();

            Collections.swap(shoppingListItems, fromPosition, toPosition);
            adapter.notifyItemMoved(fromPosition, toPosition);

            return true;
        }

        @Override
        public void onSwiped(@NonNull ViewHolder viewHolder, int direction) {
        }

        @Override
        public void onSelectedChanged(@Nullable ViewHolder viewHolder, int actionState) {
            super.onSelectedChanged(viewHolder, actionState);

            if (actionState == ACTION_STATE_DRAG) {
                viewHolder.itemView.setAlpha(0.8f);
            }
        }

        @Override
        public void clearView(@NonNull RecyclerView recyclerView, @NonNull ViewHolder viewHolder) {
            super.clearView(recyclerView, viewHolder);
            viewHolder.itemView.setAlpha(1.0f);
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_shopping_list);
        setupRecyclerView();
    }

    private void setupRecyclerView() {
        RecyclerView shoppingListRecView = findViewById(R.id.shopping_list_rec_view);

        recViewAdapter = new ShoppingListRecListAdapter(this);
        shoppingListRecView.setAdapter(recViewAdapter);

        new ItemTouchHelper(simpleTouchCallback).attachToRecyclerView(shoppingListRecView);

        ArrayList<ShoppingListItem> savedShoppingList = getSavedShoppingList();
        if (savedShoppingList != null) {
            this.shoppingListItems = savedShoppingList;
            recViewAdapter.setItems(shoppingListItems);
        }
    }

    public void onAddActionButtonPressed(View view) {
        Intent intent = new Intent(this, ShoppingItemFormActivity.class);
        startActivityForResult(intent, ADD_ITEM);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK && data != null) {
            ShoppingListItem resultItem = (ShoppingListItem) data.getSerializableExtra("ITEM");

            if (resultItem == null) {
                return;
            }

            if (requestCode == ADD_ITEM) {
                shoppingListItems.add(resultItem);
            } else if (requestCode == EDIT_ITEM) {
                int itemIndex = data.getIntExtra("ITEM_ID", -1);
                if (itemIndex != -1) {
                    shoppingListItems.set(itemIndex, resultItem);
                }
            }

            recViewAdapter.setItems(shoppingListItems);
            saveShoppingList(shoppingListItems);
        }
    }

    @Override
    public void onShoppingListItemClick(ShoppingListItem item) {
        Intent intent = new Intent(this, ShoppingItemFormActivity.class);
        intent.putExtra("ITEM", item);
        intent.putExtra("ITEM_ID", shoppingListItems.indexOf(item));
        startActivityForResult(intent, EDIT_ITEM);
    }

    private ArrayList<ShoppingListItem> getSavedShoppingList() {
        SharedPreferences sharedPreferences = getPreferences(Context.MODE_PRIVATE);
        String savedJSON = sharedPreferences.getString("shopping_list", "");
        Gson gson = new Gson();
        Type shoppingListType = new TypeToken<ArrayList<ShoppingListItem>>() {
        }.getType();
        return gson.fromJson(savedJSON, shoppingListType);
    }

    private void saveShoppingList(ArrayList<ShoppingListItem> actualShoppingList) {
        SharedPreferences sharedPreferences = getPreferences(Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        Gson gson = new Gson();
        String shoppingListJSON = gson.toJson(actualShoppingList);
        editor.putString("shopping_list", shoppingListJSON);
        editor.apply();
    }
}

File: app/src/main/java/com/mathocosta/learnbasic/activities/ShoppingListActivity.kt
package com.mathocosta.learnbasic.activities

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.mathocosta.learnbasic.R
import com.mathocosta.learnbasic.activities.adapters.ShoppingListRecListAdapter
import com.mathocosta.learnbasic.activities.adapters.ShoppingListRecListAdapter.ItemClickListener
import com.mathocosta.learnbasic.databinding.ActivityShoppingListBinding
import com.mathocosta.learnbasic.models.ShoppingList
import com.mathocosta.learnbasic.models.ShoppingListItem
import java.util.*
import kotlin.collections.ArrayList

class ShoppingListActivity : AppCompatActivity(), ItemClickListener {

    private lateinit var recViewAdapter: ShoppingListRecListAdapter
    private lateinit var binding: ActivityShoppingListBinding
    private var actionMode: ActionMode? = null

    private val simpleTouchCallback: ItemTouchHelper.SimpleCallback = object :
            ItemTouchHelper.SimpleCallback(
                    ItemTouchHelper.DOWN or ItemTouchHelper.UP, 0
            ) {
        override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
        ): Boolean = recViewAdapter.onItemMoved(viewHolder.adapterPosition, target.adapterPosition)

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {}

        override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
            super.onSelectedChanged(viewHolder, actionState)
            if (actionState == ItemTouchHelper.ACTION_STATE_DRAG) {
                viewHolder?.itemView?.alpha = 0.8f
            }
        }

        override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
            super.clearView(recyclerView, viewHolder)
            viewHolder.itemView.alpha = 1.0f
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityShoppingListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupRecyclerView()
        displayShoppingList()

        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    private fun setupRecyclerView() {
        recViewAdapter = ShoppingListRecListAdapter(this)
        binding.shoppingListRecView.adapter = recViewAdapter

        ItemTouchHelper(simpleTouchCallback).attachToRecyclerView(binding.shoppingListRecView)
    }

    private fun displayShoppingList() {
        val shoppingList = intent.getSerializableExtra("SHOPPING_LIST") as? ShoppingList ?: return
        recViewAdapter.apply {
            shoppingListItems = shoppingList.items
            notifyDataSetChanged()
        }
    }

    fun onAddActionButtonPressed(view: View?) {
        val intent = Intent(this, ShoppingItemFormActivity::class.java)
        startActivityForResult(intent, ADD_ITEM)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (resultCode == RESULT_OK && data != null) {
            val resultItem = data.getSerializableExtra("ITEM") as ShoppingListItem? ?: return

            if (requestCode == ADD_ITEM) {
                recViewAdapter.shoppingListItems.add(resultItem)
            } else if (requestCode == EDIT_ITEM) {
                val itemIndex = data.getIntExtra("ITEM_ID", -1)
                if (itemIndex != -1) {
                    recViewAdapter.shoppingListItems[itemIndex] = resultItem
                }
            }

            recViewAdapter.notifyDataSetChanged()
        }
    }

    override fun onShoppingListItemClick(item: ShoppingListItem) {
        val intent = Intent(this, ShoppingItemFormActivity::class.java).apply {
            putExtra("ITEM", item)
            putExtra("ITEM_ID", recViewAdapter.shoppingListItems.indexOf(item))
        }

        startActivityForResult(intent, EDIT_ITEM)
    }

    override fun onShoppingListItemLongClick(item: ShoppingListItem) {
        if (actionMode == null) {
            actionMode = startSupportActionMode(object : ActionMode.Callback {
                override fun onCreateActionMode(mode: ActionMode?, menu: Menu?): Boolean {
                    mode?.menuInflater?.inflate(R.menu.menu_delete, menu)
                    return true
                }

                override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?): Boolean {
                    return false
                }

                override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?): Boolean {
                    if (item?.itemId == R.id.action_delete) {
                        mode?.finish()
                        return true
                    }

                    return false
                }

                override fun onDestroyActionMode(mode: ActionMode?) {
                    actionMode = null
                }
            })
        }
    }

    companion object {
        private const val ADD_ITEM = 1
        private const val EDIT_ITEM = 2
    }
}